### ISSUE TITLE
pid1 segfault after restart, possibly when unable to unmount something?

### ISSUE THREAD

**Issue by martinetd** [systemd/systemd#21476]
**systemd version the issue has been seen with**
249.5 with a few patches: https://github.com/NixOS/nixpkgs/blob/715f63411952c86c8f57ab9e3e3cb866a015b5f2/pkgs/os-specific/linux/systemd/default.nix

**Used distribution**
nixos (downstream issue https://github.com/NixOS/nixpkgs/issues/147103 also just created)

**Linux kernel version used** (`uname -a`)
was `5.10.51` when system crashed

**CPU architecture issue was seen on**
x86_64

**Expected behaviour you didn't see**
no segfault :)

**Unexpected behaviour you saw**
segfault

**Steps to reproduce the problem**
That's a hard one, this happened when upgrading from systemd 247.6, after the upgrade (so the new one segfaulted and the issue is still relevant, but it's possible this was due to improper serialization of the older process)

**Additional program output to the terminal or log subsystem illustrating the issue**

journal when the segfault happened
```text
Nov 23 16:27:25 myhostname systemd[1]: run-media-myuser-DVD.mount: Mount process exited, code=exited, status=32/n/a
Nov 23 16:27:25 myhostname systemd[1]: Failed unmounting /run/media/myuser/DVD.
Nov 23 16:27:25 myhostname systemd[1]: Unmounting /run/media/myuser/DVD...
Nov 23 16:27:25 myhostname umount[1629960]: umount: /run/media/myuser/DVD: target is busy.
Nov 23 16:27:25 myhostname systemd[1]: run-media-myuser-DVD.mount: Mount process exited, code=exited, status=32/n/a
Nov 23 16:27:25 myhostname systemd[1]: Failed unmounting /run/media/myuser/DVD.
Nov 23 16:27:25 myhostname systemd[1]: Unmounting /run/media/myuser/DVD...
Nov 23 16:27:25 myhostname umount[1629961]: umount: /run/media/myuser/DVD: target is busy.
Nov 23 16:27:25 myhostname systemd[1]: run-media-myuser-DVD.mount: Mount process exited, code=exited, status=32/n/a
Nov 23 16:27:25 myhostname systemd[1]: Failed unmounting /run/media/myuser/DVD.
Nov 23 16:27:25 myhostname systemd[1]: Unmounting /run/media/myuser/DVD...
Nov 23 16:27:25 myhostname umount[1629964]: umount: /run/media/myuser/DVD: target is busy.
Nov 23 16:27:25 myhostname systemd[1]: run-media-myuser-DVD.mount: Mount process exited, code=exited, status=32/n/a
Nov 23 16:27:25 myhostname systemd[1]: Failed unmounting /run/media/myuser/DVD.
Nov 23 16:27:25 myhostname systemd[1]: Unmounting /run/media/myuser/DVD...
Nov 23 16:27:25 myhostname umount[1629965]: umount: /run/media/myuser/DVD: target is busy.
Nov 23 16:27:25 myhostname systemd[1]: run-media-myuser-DVD.mount: Mount process exited, code=exited, status=32/n/a
Nov 23 16:27:25 myhostname systemd[1]: Failed unmounting /run/media/myuser/DVD.
Nov 23 16:27:25 myhostname systemd[1]: Unmounting /run/media/myuser/DVD...
Nov 23 16:27:25 myhostname umount[1629966]: umount: /run/media/myuser/DVD: target is busy.
Nov 23 16:27:25 myhostname systemd[1]: run-media-myuser-DVD.mount: Mount process exited, code=exited, status=32/n/a
Nov 23 16:27:25 myhostname systemd[1]: Failed unmounting /run/media/myuser/DVD.
Nov 23 16:27:25 myhostname kernel: show_signal_msg: 814 callbacks suppressed
Nov 23 16:27:25 myhostname kernel: systemd[1]: segfault at 18 ip 00005882524303bf sp 00007ffe9f5c3858 error 4 in systemd[5882523d5000+dc000]
Nov 23 16:27:25 myhostname kernel: Code: 5a 49 8b 04 24 48 8d 35 bf 8e 09 00 b9 69 05 00 00 4c 8d 05 13 a7 09 00 4c 8b 88 20 06 00 00 57 bf 04 00 00 00 48 8b 54 24 20 <ff> 72>
Nov 23 16:27:25 myhostname audit[1629970]: ANOM_ABEND auid=4294967295 uid=0 gid=0 ses=4294967295 subj=kernel pid=1629970 comm="systemd" exe="/nix/store/cfqfzy5k6l7b2f3bxy2zdh>
Nov 23 16:27:25 myhostname kernel: audit: type=1701 audit(1637652445.505:1149): auid=4294967295 uid=0 gid=0 ses=4294967295 subj=kernel pid=1629970 comm="systemd" exe="/nix/st>
Nov 23 16:27:25 myhostname systemd-coredump[1629971]: Due to PID 1 having crashed coredump collection will now be turned off.
Nov 23 16:27:25 myhostname systemd-coredump[1629971]: [🡕] Process 1629970 (systemd) of user 0 dumped core.
Nov 23 16:27:25 myhostname systemd[1]: Caught <SEGV>, dumped core as pid 1629970.
Nov 23 16:27:25 myhostname kernel: audit: type=1334 audit(1637652445.619:1150): prog-id=68026 op=UNLOAD
Nov 23 16:27:25 myhostname kernel: audit: type=1334 audit(1637652445.619:1151): prog-id=68025 op=UNLOAD
Nov 23 16:27:25 myhostname audit: BPF prog-id=68026 op=UNLOAD
Nov 23 16:27:25 myhostname audit: BPF prog-id=68025 op=UNLOAD
Nov 23 16:27:25 myhostname systemd[1]: Freezing execution.
```

coredumpctl gdb and backtrace, but unfortunately I have no symbols available
```text
Reading symbols from /nix/store/cfqfzy5k6l7b2f3bxy2zdhp9sjiw8ijd-systemd-249.5/lib/systemd/systemd...
(No debugging symbols found in /nix/store/cfqfzy5k6l7b2f3bxy2zdhp9sjiw8ijd-systemd-249.5/lib/systemd/systemd)
[New LWP 1629970]
[New LWP 1]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/nix/store/ff88p8pnhdmf8bflzbxldys21djw9dp0-glibc-2.33-56/lib/libthread_db.so.1".
Core was generated by `/run/current-system/systemd/lib/systemd/systemd --system --deserialize 275'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x000079d9187e8e27 in kill () from /nix/store/ff88p8pnhdmf8bflzbxldys21djw9dp0-glibc-2.33-56/lib/libc.so.6
[Current thread is 1 (LWP 1629970)]
(gdb) bt
#0  0x000079d9187e8e27 in kill () from /nix/store/ff88p8pnhdmf8bflzbxldys21djw9dp0-glibc-2.33-56/lib/libc.so.6
#1  0x00005882523df219 in crash ()
#2  <signal handler called>
#3  0x00005882524303bf in manager_loop ()
#4  0x00005882523da237 in main ()
(gdb) disass manager_loop
... 
   0x000058825243036c <+2332>:	jmp    0x5882524303f8 <manager_loop+2472>
   0x0000588252430371 <+2337>:	call   0x5882523d6410 <log_get_max_level@plt>
   0x0000588252430376 <+2342>:	lea    0x40(%rsp),%r13
   0x000058825243037b <+2347>:	cmp    $0x3,%eax
   0x000058825243037e <+2350>:	jle    0x5882524303f0 <manager_loop+2464>
   0x0000588252430380 <+2352>:	mov    %r12,%rdi
   0x0000588252430383 <+2355>:	call   0x58825246d6e0 <unit_get_exec_context>
   0x0000588252430388 <+2360>:	test   %rax,%rax
   0x000058825243038b <+2363>:	je     0x588252430396 <manager_loop+2374>
   0x000058825243038d <+2365>:	cmpl   $0x3,0x2b4(%rax)
   0x0000588252430394 <+2372>:	jbe    0x5882524303f0 <manager_loop+2464>
   0x0000588252430396 <+2374>:	mov    (%r12),%rax
   0x000058825243039a <+2378>:	lea    0x98ebf(%rip),%rsi        # 0x5882524c9260
   0x00005882524303a1 <+2385>:	mov    $0x569,%ecx
   0x00005882524303a6 <+2390>:	lea    0x9a713(%rip),%r8        # 0x5882524caac0 <__func__.80>
   0x00005882524303ad <+2397>:	mov    0x620(%rax),%r9
   0x00005882524303b4 <+2404>:	push   %rdi
   0x00005882524303b5 <+2405>:	mov    $0x4,%edi
   0x00005882524303ba <+2410>:	mov    0x20(%rsp),%rdx
=> 0x00005882524303bf <+2415>:	push   0x18(%rdx)
(gdb) info reg
rax            0x0                 0
rbx            0x7ffe9f5c30e8      140731572039912
rcx            0x79d9187e8e27      133973325811239
rdx            0x79d918899b67      133973326535527
rsi            0xb                 11
rdi            0x18df12            1629970
rbp            0x7ffe9f5c30e0      0x7ffe9f5c30e0
rsp            0x7ffe9f5c3038      0x7ffe9f5c3038
r8             0x7ffe9f5c3050      140731572039760
r9             0x588253666d50      97316768214352
r10            0x4                 4
r11            0x207               519
r12            0xb                 11
r13            0x0                 0
r14            0x0                 0
r15            0x1                 1
rip            0x79d9187e8e27      0x79d9187e8e27 <kill+7>
eflags         0x207               [ CF PF IF ]
cs             0x33                51
ss             0x2b                43
ds             0x0                 0
es             0x0                 0
fs             0x0                 0
gs             0x0                 0
(gdb) x/gx $rdx
0x79d918899b67 <chdir+7>:	0x0173fffff0013d48

```

**Comment by poettering** [comment]
Please provide full backtrace with debug symbols, otherwise this is not actionable for us!

**Comment by martinetd** [comment]
I've asked if it's possible to get debug symbols for existing nixos binaries, but I'm not sure it's possible.
As I can't reproduce this (only happened once, no idea if the unmount problem has anything to do with it) it might be a bit of time until I can provide a full backtrace, I'll spend a bit more time looking at the assembly to identify what variable this is and where in the function instead if it doesn't look like anything already known.

**Comment by martinetd** [comment]
Rebuilding with the exact same option apparently made gdb happy, here's a full backtrace:
```
(gdb) bt full
#0  0x000079d9187e8e27 in kill () from /nix/store/ff88p8pnhdmf8bflzbxldys21djw9dp0-glibc-2.33-56/lib/libc.so.6
No symbol table info available.
#1  0x00005882523df219 in crash (sig=11) at ../src/core/main.c:257
        sa = {__sigaction_handler = {sa_handler = 0x0, sa_sigaction = 0x0}, sa_mask = {__val = {0 <repeats 16 times>}}, sa_flags = 0, sa_restorer = 0x0}
        pid = <optimized out>
        __PRETTY_FUNCTION__ = "crash"
        __func__ = "crash"
#2  <signal handler called>
No symbol table info available.
#3  0x00005882524303bf in manager_dispatch_stop_when_bound_queue (m=<optimized out>) at ../src/core/manager.c:1385
        _u = 0x588253741490
        _l = 4
        error = {name = 0x0, message = 0x0, _need_free = 0}
        culprit = 0x0
        n = 1
        u = 0x588253741490
        r = <optimized out>
        n = <optimized out>
        u = <optimized out>
        r = <optimized out>
        __PRETTY_FUNCTION__ = "manager_dispatch_stop_when_bound_queue"
        __func__ = "manager_dispatch_stop_when_bound_queue"
        error = <optimized out>
        culprit = <optimized out>
        _head = <optimized out>
        _item = <optimized out>
        _u = <optimized out>
        _l = <optimized out>
        _u = <optimized out>
        _l = <optimized out>
        _error = <optimized out>
        _u = <optimized out>
        _l = <optimized out>
#4  manager_loop (m=m@entry=0x5882536386e0) at ../src/core/manager.c:2983
        wait_usec = <optimized out>
        watchdog_usec = <optimized out>
        rl = {interval = 1000000, burst = 50000, num = 362, begin = 9963679270635}
        r = <optimized out>
        __PRETTY_FUNCTION__ = "manager_loop"
        __func__ = "manager_loop"
#5  0x00005882523da237 in invoke_main_loop (ret_error_message=0x7ffe9f5c3938, ret_switch_root_init=<synthetic pointer>, ret_switch_root_dir=<synthetic pointer>, 
    ret_fds=0x7ffe9f5c3948, ret_shutdown_verb=<synthetic pointer>, ret_retval=<synthetic pointer>, ret_reexecute=<synthetic pointer>, 
    saved_rlimit_memlock=0x7ffe9f5c39b0, saved_rlimit_nofile=0x7ffe9f5c39a0, m=0x5882536386e0) at ../src/core/main.c:1916
        r = <optimized out>
        r = <optimized out>
        __PRETTY_FUNCTION__ = "invoke_main_loop"
        __func__ = "invoke_main_loop"
--Type <RET> for more, q to quit, c to continue without paging--
        _error = <optimized out>
        _level = <optimized out>
        _e = <optimized out>
        saved_log_target = <optimized out>
        saved_log_level = <optimized out>
        _level = <optimized out>
        _e = <optimized out>
        _level = <optimized out>
        _e = <optimized out>
        _level = <optimized out>
        _e = <optimized out>
        _ptr_ = <optimized out>
        _ptr_ = <optimized out>
        _level = <optimized out>
        _e = <optimized out>
        table = {0x0, 0x5882524ce2d0 "exit", 0x0, 0x0, 0x5882524b1a6c "reboot", 0x5882524b4872 "poweroff", 0x5882524b487b "halt", 0x5882524b35a2 "kexec", 0x0}
        _level = <optimized out>
        _e = <optimized out>
#6  main (argc=4, argv=0x7ffe9f5c3c98) at ../src/core/main.c:2927
        initrd_timestamp = <optimized out>
        userspace_timestamp = {realtime = 1637652443919708, monotonic = 9963678034157}
        kernel_timestamp = {realtime = 1627688765885550, monotonic = 0}
        security_start_timestamp = {realtime = 0, monotonic = 0}
        security_finish_timestamp = {realtime = 0, monotonic = 0}
        saved_rlimit_nofile = {rlim_cur = 1024, rlim_max = 4096}
        saved_rlimit_memlock = {rlim_cur = 65536, rlim_max = 65536}
        skip_setup = <optimized out>
        loaded_policy = false
        queue_default_job = <optimized out>
        first_boot = <optimized out>
        reexecute = false
        switch_root_dir = 0x0
        switch_root_init = 0x0
        before_startup = 9963678058335
        after_startup = <optimized out>
        systemd = "systemd"
        timespan = "\000\000\000\000\000\000\000\000K\000\000\000\000\000\000\000\004\000\000\000\000\000\000\000\330\376\377\377\377\377\377\377\370;\\\237\376\177\000\000\000\000\000\000\000\000\000\000H\f\227\030\331y\000\000`\255\326\030\331y\000"
        shutdown_verb = 0x0
        error_message = 0x0
        r = <optimized out>
        retval = 1
        m = 0x5882536386e0
        fds = 0x0
        __PRETTY_FUNCTION__ = "main"
        __func__ = "main"
(gdb) list manager_dispatch_stop_when_bound_queue
... 
1375	                if (!unit_is_bound_by_inactive(u, &culprit))
1376	                        continue;
1377	
1378	                log_unit_debug(u, "Unit is stopped because bound to inactive unit %s.", culprit->id);
1379	
1380	                /* If stopping a unit fails continuously we might enter a stop loop here, hence stop acting on the
1381	                 * service being unnecessary after a while. */
1382	
1383	                if (!ratelimit_below(&u->auto_start_stop_ratelimit)) {
1384	                        log_unit_warning(u, "Unit needs to be stopped because it is bound to inactive unit %s it, but not stopping since we tried this too often recently.", culprit->id);
1385	                        continue;
1386	                }
1387	
1388	                r = manager_add_job(u->manager, JOB_STOP, u, JOB_REPLACE, NULL, &error, NULL);
1389	                if (r < 0)
(gdb) p culprit
$1 = (Unit *) 0x0
(gdb) p u->id
$2 = 0x58825364d340 "run-media-myuser-DVD.mount"
```

(so this confirms my suspicion that this was caused by the mount point that failed to unmount too many times...)

Looking at the code here `unit_is_bound_by_inactive` (`unit_is_upheld_by_active` as of today's main branch) should always set culprit if it returned true, and if we got there it did return true, and `other` in that function definitely couldn't have been NULL, so I'm a bit perplexed, but the string constants surrounding the fault address in the disassembler output match (`Unit needs to be stopped because it is bound to inactive unit %s it, but not stopping since we tried this too often recently.` among others) and `0x18(%rdx)` offset corresponds to Unit's `->id` and `$rdx` = 0, so there is no doubt about it.
I'm not sure what to suggest here, so I'll leave the rest to you for now...

**Comment by martinetd** [comment]
Since I don't like not understanding, I double-checked the assembly I have... Feel free to ignore this message. Full assembly for the function can be found here: https://gaia.codewreck.org/local/tmp/systemd_disass.txt (can't figure how to attach a text file here? it only lets me attach images.. I'll eventually remove that file so if someone could attach it here for archival that'd be great)

So we're calling (after macro expansions) `log_object_internal` with these arguments, passed as follow:
```
log_object_internal(_l, error, PROJECT_FILE, __LINE__, __func__, _u->manager->unit_log_field, _u->id, _u->manager->invocation_log_field, _u->invocation_id_string, ##__VA_ARGS__)
 (with va_args = "unit needs to be stopped..." + culprit->id)
RDI = _l (log level)
RSI = error
RDX = PROJECT_FILE
RCX = __LINE__
R8 = __func__
R9 = u->manager->unit_log_field
stack1 u->id
stack2 u->manager->invocation_log_field,
stack3 u->invocation_id_string
stack4 "unit needs to be stopped..."
stack5 culprit->id
```

Which allows annotating the assembly as follow
```
   0x000058825243039a <+2378>:	lea    0x98ebf(%rip),%rsi        # 0x5882524c9260 <- "Unit needs to be stopped..." message, pushed later
   0x00005882524303a1 <+2385>:	mov    $0x569,%ecx <- __LINE__
   0x00005882524303a6 <+2390>:	lea    0x9a713(%rip),%r8        # 0x5882524caac0  <__func__.80> <- __func__
   0x00005882524303ad <+2397>:	mov    0x620(%rax),%r9 <- u->manager->unit_log_field
   0x00005882524303b4 <+2404>:	push   %rdi <- (hm it's just saving the previous value of rdi before overwriting it to get it back later?)
   0x00005882524303b5 <+2405>:	mov    $0x4,%edi <- level=warn
   0x00005882524303ba <+2410>:	mov    0x20(%rsp),%rdx <- would be culprit?!
=> 0x00005882524303bf <+2415>:	push   0x18(%rdx) <-culprit->id
   0x00005882524303c2 <+2418>:	lea    0x3e0(%r12),%rdx <- u->invocation_id_string, pushed later
   0x00005882524303ca <+2426>:	push   %rsi <- "Unit needs to be stopped..." message
   0x00005882524303cb <+2427>:	xor    %esi,%esi <- error = 0
   0x00005882524303cd <+2429>:	push   %rdx <- u->invocation_id_string
   0x00005882524303ce <+2430>:	lea    0x96d59(%rip),%rdx        # 0x5882524c712e <- PROJECT_FILE
   0x00005882524303d5 <+2437>:	push   0x630(%rax) <- u->manager->invocation_log_field,
   0x00005882524303db <+2443>:	xor    %eax,%eax
   0x00005882524303dd <+2445>:	push   0x18(%r12) <- u->id
   0x00005882524303e2 <+2450>:	call   0x5882523d8860 <log_object_internal@plt>
```

The question is culprit apparently comes from the stack... culprit has been written a bit down below
```
   0x000058825243047f <+2607>:  mov    (%rsp),%rsi
   0x0000588252430483 <+2611>:  mov    %r12,%rdi
...
   0x00005882524304a2 <+2642>:  call   0x588252468f80 <unit_is_bound_by_inactive>
   0x00005882524304a7 <+2647>:  test   %al,%al
   0x00005882524304a9 <+2649>:  je     0x5882524303eb <manager_loop+2459>
```
So the address passed for &culprit was whatever was at the top of the stack, and it seems to match:
```
(gdb) x/gx $rsp+0x8 (we had one push earlier, so %rsp at this point is %rsp+8 now)
0x7ffe9f5c3860:	0x00007ffe9f5c3878
(gdb) x/gx $rsp+0x20
0x7ffe9f5c3878:	0x0000000000000000 <- $rsp+0x20 we're accessing now is the same address alright
```

As the function must have set it earlier, the only possibly I see is something else in between having overwritten the value? But optimizations make the code flow hard to follow and I ran out of time for tonight; will continue later unless I got beaten to it.

**Comment by poettering** [comment]
any chance you can rebuild/reproduce with optimizations off?

**Comment by arianvp** [comment]
@martinetd 

To disable optimizations and enable debug symbols you can add this to your nixos config:

```
{pkgs,config,...}: {
  systemd.package = pkgs.enableDebugging pkgs.systemd;
}
```

This will require a rebuild though. But given this only replaces the binaries of systemd; and not e.g. `libsystemd` it should be pretty quick (will not cascade a mass rebuild).

**Comment by martinetd** [comment]
Rebuilding with symbols is easy, I'm just not sure I can reproduce this easily.

I mean, I could switch back to the old configuration, remount a DVD, make it used etc and maybe it'll reproduce but this is not a system I'm comfortable rebooting all the time.

I'll first see if I can reproduce with the exact same binaries in a VM by switching back and forth in a somewhat similar environment, and if I can I'll try again without optimizations, but it might take a bit of time.

**Comment by martinetd** [comment]
Well, there's some good news:
 - I could reproduce in the VM!
 - enableDebugging is broken on my commit: https://github.com/NixOS/nixpkgs/issues/136756 -- not that it's difficult to do otherwise (pkgs.systemd.overrideAttrs...)
 -  Ultimately I was too confident that unit_is_bound_by_inactive() was correct and missed a typo: the debug build didn't really help with that, but a second look did wonders... Since I was able to reproduce quite reliably I've just finished rebuilding with a patch similar to the one in the PR, and can now finally see the correct message: the culprit was dev-sr0.device! murder scene can be closed :)
